runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1086 @@
+			}
+
+			Self::deposit_event(Event::<T>::TemporarySlotAssigned(id));
+
+			Ok(())
+		}
+
+		/// Unassign a permanent or temporary parachain slot
+		#[pallet::weight((1_000, DispatchClass::Operational))]
+		pub fn unassign_parachain_slot(origin: OriginFor<T>, id: ParaId) -> DispatchResult {
+			T::AssignSlotOrigin::ensure_origin(origin.clone())?;
+
+			ensure!(
+				Self::has_permanent_slot(id) || Self::has_temporary_slot(id),
+				Error::<T>::SlotNotAssigned
+			);
+
+			if PermanentSlots::<T>::contains_key(id) {
+				<PermanentSlotCount<T>>::try_mutate(|count| -> DispatchResult {
+					*count = count.checked_sub(1).ok_or(ArithmeticError::Underflow)?;
+					Ok(())
+				})?;
+				PermanentSlots::<T>::remove(id);
+			} else if TemporarySlots::<T>::contains_key(id) {
+				TemporarySlots::<T>::remove(id);
+			}
+
+			// Clean any para lease
+			Self::clear_slot_leases(origin.clone(), id)?;
+
+			// Force downgrade to parathread (if needed) before end of lease period
+			if paras::Pallet::<T>::lifecycle(id) == Some(ParaLifecycle::Parachain) {
+				runtime_parachains::schedule_parachain_downgrade::<T>(id)
+					.map_err(|_| Error::<T>::CannotDowngrade)?;
+			}
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn allocate_temporary_slot_leases(lease_period_index: LeasePeriodOf<T>) -> DispatchResult {
+		let mut active_temp_slots = 0u32;
+		let mut pending_temp_slots = Vec::new();
+		TemporarySlots::<T>::iter().for_each(|(para, maybe_slot)| {
+			if let Some(slot) = maybe_slot {
+				match slot.last_lease {
+					Some(last_lease)
+						if last_lease <= lease_period_index &&
+							lease_period_index <
+								(last_lease.saturating_add(slot.period_count)) =>
+					{
+						// Active slot lease
+						active_temp_slots += 1;
+					}
+					Some(last_lease)
+						// Slot w/ past lease, only consider it every other slot lease period (times period_count)
+						if last_lease.saturating_add(slot.period_count.saturating_mul(2u32.into())) <= lease_period_index => {
+							pending_temp_slots.push((para, slot));
+					},
+					None if slot.period_begin <= lease_period_index => {
+						// Slot hasn't had a lease yet
+						pending_temp_slots.insert(0, (para, slot));
+					},
+					_ => {
+						// Slot not being considered for this lease period (will be for a subsequent one)
+					},
+				}
+			}
+		});
+
+		if active_temp_slots < T::MaxTemporarySlotPerLeasePeriod::get() &&
+			!pending_temp_slots.is_empty()
+		{
+			// Sort by lease_count, favoring slots that had no or less turns first
+			// (then by last_lease index, and then Para ID)
+			pending_temp_slots.sort_by(|a, b| {
+				a.1.lease_count
+					.cmp(&b.1.lease_count)
+					.then_with(|| a.1.last_lease.cmp(&b.1.last_lease))
+					.then_with(|| a.0.cmp(&b.0))
+			});
+
+			let slots_to_be_upgraded = pending_temp_slots
+				.iter()
+				.take((T::MaxTemporarySlotPerLeasePeriod::get() - active_temp_slots) as usize)
+				.collect::<Vec<_>>();
```

Comment:
this collect seems unnecessary to me, you just iterate by reference anyway

Reply:
Not a biggie I think.

Reply:
sure :shrug:

Reply:
But you're right, I did the change.